### ISSUE TITLE
Support operations in autograd tensor that are needed by convolutional layer and cross entropy

### ISSUE THREAD

**Issue by iamtrask** [OpenMined/PySyft#3509]
From @vvmnnnkv:

> when we perform computations using custom tensors (like Autograd, FixedPrecision, AdditiveSharing) must all those computations be trivial like add/mul/exp/etc?
> For example, if we need torch.softmax(autograd_tensor), it is required to implement our own syft.softmax() that does exp(x)/exp(x).sum() and use it instead?
> Specifically for Autograd tensor, it is seems possible to write grad_fn for any "high-level" torch operation, like torch.softmax, F.cross_entropy, F.conv2d, etc. but this won't probably work together with other custom tensor types, e.g. when you'll have AT>FPT>AST chain, so this way is not desirable?
> Why asking: approach of re-writing torch functions in python to break them to basic ops seems very limiting for FL.
> I thought the premise was like "make the model with pytorch, as usual, and host it for FL". The reality will be "make the model with syft or hand-written components...". Which means implementing a tricky model might be as hard as the FL project itself

> It seems to be possible to implement autograd tracing simpler way, by overriding torch.nn.functional.cross_entropy(probs, targets) in Autograd tensor (assuming here that probs is AT), where we could pass-through args to native torch method, and add backward grad calculation (which can be as simple as probs - targets).
> But other tensors like FPT, AST won't be able to do same as they will need special cross_entropy broken down to basic ops that they can override, so there will be kind of inconsistency in what tensors types can support and there will be cases when you can't put them in chain.

> At very high-level it seems there're 2 options to make autograd tracing:
> a) write custom gradfn for torch function
> b) implement torch function in python using basic operations (and have custom gradfn for all these operations)
> Both feel bad, at least for FL purpose. But maybe we don't have other choice.
> Option b) is worse in terms or performance.
